### PR TITLE
scheduler: warn when system jobs cannot place an alloc

### PR DESCRIPTION
**Pull request by notnoop**
When a system or sysbatch job specify constraints that none of the current nodes meet, report a warning to the user.

Also, for sysbatch job, mark the job as dead if no allocation is placed at all. System jobs behavior isn't affected by this.

A sample run would look like:

```
$ nomad job run ./example.nomad
==> 2021-08-31T16:57:35-04:00: Monitoring evaluation "b48e8882"
    2021-08-31T16:57:35-04:00: Evaluation triggered by job "example"
==> 2021-08-31T16:57:36-04:00: Monitoring evaluation "b48e8882"
    2021-08-31T16:57:36-04:00: Evaluation status changed: "pending" -> "complete"
==> 2021-08-31T16:57:36-04:00: Evaluation "b48e8882" finished with status "complete" but failed to place all allocations:
    2021-08-31T16:57:36-04:00: Task Group "cache" (failed to place 1 allocation):
      * Constraint "${meta.tag} = bar": 2 nodes excluded by filter
      * Constraint "${attr.kernel.name} = linux": 1 nodes excluded by filter

$ nomad job status example
ID            = example
Name          = example
Submit Date   = 2021-08-31T16:57:35-04:00
Type          = sysbatch
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = dead
Periodic      = false
Parameterized = false

Summary
Task Group  Queued  Starting  Running  Failed  Complete  Lost
cache       0       0         0        0       0         0

Allocations
No allocations placed
```

### CODE REVIEW

File: scheduler/scheduler_sysbatch_test.go [hunk 1]
```diff
@@ -890,8 +945,8 @@ func TestSysBatch_JobConstraint_AddNode(t *testing.T) {
 	// Add a new node Class-B
 	var nodeBTwo *structs.Node
 	nodeBTwo = mock.Node()
-	require.NoError(t, nodeBTwo.ComputeClass())
 	nodeBTwo.NodeClass = "Class-B"
+	require.NoError(t, nodeBTwo.ComputeClass())
```
Comment (notnoop) [inline scheduler/scheduler_sysbatch_test.go:949]: Good catch!

File: scheduler/scheduler_sysbatch_test.go [hunk 2]
```diff
@@ -910,8 +965,10 @@ func TestSysBatch_JobConstraint_AddNode(t *testing.T) {
 	require.Nil(t, h.Process(NewSysBatchScheduler, eval3))
 	require.Equal(t, "complete", h.Evals[2].Status)
 
-	// Ensure no failed TG allocs
-	require.Equal(t, 0, len(h.Evals[2].FailedTGAllocs))
+	// Ensure `groupA` fails to be placed due to its constraint, but `groupB` doesn't
+	require.Equal(t, 1, len(h.Evals[2].FailedTGAllocs))
```
Comment (notnoop) [inline scheduler/scheduler_sysbatch_test.go:969]: [`require.Len`](https://pkg.go.dev/github.com/stretchr/testify/require#Len) is a handy test for this; on failure, it prints the map so we can inspect state of map to understand how the count differed.

```suggestion
	require.Len(t, h.Evals[2].FailedTGAllocs, 1)
```

**Review by lgfa29 — APPROVED** [review]

**Comment by lgfa29** [comment]
@notnoop I pushed a changelog entry, but I'm not sure if it's good. Feel free to change it before merge.

File: scheduler/scheduler_system.go [hunk 3]
```diff
@@ -279,15 +279,38 @@ func (s *SystemScheduler) computeJobAllocs() error {
 	return s.computePlacements(diff.place)
 }
 
+func mergeNodeFiltered(acc, curr *structs.AllocMetric) *structs.AllocMetric {
+	if acc == nil {
+		v := *curr
+		return &v
```
Comment (schmichael) [inline scheduler/scheduler_system.go:285]: Shallow copying means referenced values in `curr` on the first call will be mutated on successive calls.

**Review by schmichael — CHANGES_REQUESTED** [review]
@notnoop is out, so I'll try to move this forward! I'm not sure my comment is a blocker, but I'd like to check it out before merging.

File: .changelog/11111.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+scheduler: Keep track of system and sysbatch evaluations that fail to place an allocation
```
Comment (schmichael) [inline .changelog/11111.txt:2]:
```suggestion
scheduler: Warn users when system and sysbatch evaluations that fail to place an allocation
```

While I don't think this fully encompasses the changes, other bits like marking sysbatch as `dead` immediately are new so haven't changed from a previous release.

**Review by schmichael — APPROVED** [review]

**Pull request merged**